### PR TITLE
[RESTEASY-2567] Added new test coverage for JSON merge patch

### PR DESCRIPTION
**Pull request by honza-kasik**
Mainly negative test cases, see test documentation for details.

See https://issues.redhat.com/browse/RESTEASY-2567

### CODE REVIEW

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java
Comment (jamezp) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java:1]: This needs a copyright header.

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+    @AfterEach
+    public void afterEach() {
+        deployer.undeploy(PATCH_DEPLOYMENT);
+    }
+
+    @Deployment(managed = false, name = PATCH_DEPLOYMENT)
+    public static Archive<?> createDeployment() {
+        WebArchive war = TestUtil.prepareArchive(PATCH_DEPLOYMENT);
+        return TestUtil.finishContainerPrepare(war, null, StudentResource.class, Student.class);
+    }
```
Comment (jamezp) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java:78]: Why are we not using a managed deployment?

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+public class StudentJsonMergePatchTest {
+
+    private static final Logger log = Logger.getLogger(StudentJsonMergePatchTest.class);
+
+    @ArquillianResource
+    private Deployer deployer;
+
+    static Client client;
+    static final String PATCH_DEPLOYMENT = "Patch";
+    static final Student testStudentEntity = new Student()
```
Comment (jamezp) [inline testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java:47]: Minor nit, there should be no reason for these not to be `private`.

**Review by jamezp — CHANGES_REQUESTED** [review]

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java (hunk 1 above)
Comment (honza-kasik) [reply testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java:78]: The deployment is using a `Map` to store entities. To make tests as independent as possible we need the `Map` to be empty for each test. So either we have to have a deployment for each test or reuse deployment with refreshed state. To achieve the initial state with empty `Map` I decided to use non-managed deployment with redeployment in `each` JUnit lifecycle.

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java (hunk 2 above)
Comment (honza-kasik) [reply testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java]: Thanks, fixed.

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java (hunk 1 above)
Comment (jamezp) [reply testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java:79]: Hmm... ...okay. Not suggesting we need to do this, but we could add a method to clear the map. I"m not sure if it's faster to redeploy or send a request to clear the map. 

This shouldn't block this PR though IMO.

Comment (honza-kasik) [reply testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java:79]: Thanks for the feedback. This solution in my opinion is the simplest available - we don't need to make additional checks whether the map is really empty after emptying it somehow and introduce any new methods for the deployment.

File: testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java
Comment (honza-kasik) [reply testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java:1]: Added, thanks

**Comment by jamezp** [comment]
@honza-kasik It looks like the format check failed.
```
Error:  Failed to execute goal net.revelc.code:impsort-maven-plugin:1.11.0:check (check-import-sort) on project resteasy-integration-tests: Imports are not sorted in /home/runner/work/resteasy/resteasy/testsuite/integration-tests/src/test/java/org/jboss/resteasy/test/resource/patch/StudentJsonMergePatchTest.java -> [Help 1]
```

This can be resolved by simply running `mvn clean process-test-classes`.

Also, can you squash the commits into a single commit?

**Review by jamezp — APPROVED** [review]

**Comment by jamezp** [comment]
Upstream #4383

**Pull request merged**